PythonPractice/TicTacToe/tictactoe.py

```diff
@@ -0,0 +1,155 @@
+        """
+        if self.num_turns > 0:
+            good_input = False
+            row = None
+            col = None
+            # communicate who should play next
+            print "-----------------------------"
+            print "PLAYER %s's turn" % self.player
+            # get validated user input
+            while not good_input:
+                try:
+                    row = int(raw_input("What row?\n"))-1
+                    col = int(raw_input("What column?\n"))-1
+                    if row < 0 or col < 0:
+                        raise ValueError
+                    if self.board[row][col]:
+                        raise SpaceTakenError
+                    if row > self.size or col > self.size:
+                        raise IndexError
+                except ValueError:
+                    print "Make sure the row and column are positive integers."
+                except SpaceTakenError:
+                    print "That space is already taken. Try again."
+                except IndexError:
+                    print "That space is out of bounds. Try again."
+                else:
+                    good_input = True
+            # set space with user's value (X or O) depending on whose turn it is
+            # update next player
+            if self.player == self.playerX:
+                self.board[row][col] = self.playerX
+                self.player = self.playerO
+            else:
+                self.board[row][col] = self.playerO
+                self.player = self.playerX
+            self.num_turns -= 1
+            return True
+        else:
+            return False
+
+    def check_winner(self):
+        """
+        Checks winning conditions:
+            -win along a diagonal (left to right or right to left)
+            -win along row
+            -win along column
+        Returns:
+            -winning letter
+        """
+        winner = None
+        # check top left corner to bottom right corner diagonal win
+        for i in range(self.size):
+            first_element_in_left_diag = self.board[0][0]
+            if not first_element_in_left_diag or self.board[i][i] != first_element_in_left_diag:
+                break
+        else:
+            return first_element_in_left_diag
+        # check top right corner to bottom left corner diagonal win
+        for i in range(self.size):
+            first_element_in_right_diag = self.board[0][self.size-1]
+            if not first_element_in_right_diag or self.board[i][self.size-1-i] != first_element_in_right_diag:
+                break
+        else:
+            return first_element_in_right_diag
+        # check row and column wins
+        for i in range(self.size):
+            first_element_in_row = self.board[i][0]
+            first_element_in_col = self.board[0][i]
+            # rows
+            for j in range(1, self.size):
+                if not first_element_in_row or self.board[i][j] != first_element_in_row:
```

Comment:
If first_element_in_row evaluates to True, this arrangement will cause it to be redundantly rechecked for each iteration through the loop.